### PR TITLE
[python] Add new plot type: 'constraint2D'

### PR DESCRIPTION
**Pull request by dvandyk**
@mreboud Could you please check my logic to compute the ellipses?

### CODE REVIEW

File: python/eos/plot/plotter.py [hunk 1]
```diff
@@ -671,6 +672,181 @@ def plot(self):
...
+                self.xobservable = xitem['observable']
+                self.xkinematics = xitem['kinematics'] if 'kinematics' in xitem else {}
+                self.xoptions    = xitem['options']    if 'options'    in xitem else {}
+
+                if 'y' not in item:
+                    raise KeyError('no y-axis specification for 2D constraint')
+
+                yitem = item['y']
+                if 'observable' not in yitem:
+                    raise KeyError('no observable in 2D constraint\'s x-axis specification')
```
Comment (pluegh) [inline python/eos/plot/plotter.py:721]: `x-axis` → `y-axis`

File: python/eos/plot/plotter.py [hunk 2]
```diff
@@ -671,6 +672,181 @@ def plot(self):
...
+
+                    means  = np.array(constraint['means'])
+                    xmean  = means[xidx]
+                    ymean  = means[yidx]
+
+                    cov    = np.array(constraint['covariance'])
+                    xsigma = np.sqrt(covariance[xidx, xidx])
+                    ysigma = np.sqrt(covariance[yidx, yidx])
+                    rho    = cov[xidx, yidx] / (xsigma * ysigma)
+                else:
```
Comment (pluegh) [inline python/eos/plot/plotter.py:749]: `if constraint['type'] == 'MultivariateGaussian'` is easier to understand

File: python/eos/plot/plotter.py [hunk 3]
```diff
@@ -671,6 +672,181 @@ def plot(self):
...
+                sigmas = [1.0, 2.0, 3.0]
+                alphas = np.linspace(0.0, self.alpha, len(sigmas) + 1)[1:]
+                for sigma, alpha in zip(sorted(sigmas, reverse=True), alphas):
+                    xwidth = 2.0 * xsigma * np.sqrt(1.0 + rho) * sigma
+                    ywidth = 2.0 * ysigma * np.sqrt(1.0 - rho) * sigma
+
+                    ellipse = matplotlib.patches.Ellipse((xmean, ymean), width=xwidth, height=ywidth, angle=45,
+                                          alpha=alpha, color=self.color, linewidth=1, fill=True)
+                    self.plotter.ax.add_patch(ellipse)
+            else: # univariate gaussian
```
Comment (pluegh) [inline python/eos/plot/plotter.py:786]: again, `if constraint['type'] == 'Gaussian'` would be easier to comprehend

File: python/eos/plot/plotter.py [hunk 4]
```diff
@@ -671,6 +672,181 @@ def plot(self):
...
+                    sigma_hi = np.sqrt(constraint['sigma-stat']['hi']**2 + constraint['sigma-sys']['hi']**2)
+                    sigma_lo = np.sqrt(constraint['sigma-stat']['lo']**2 + constraint['sigma-sys']['lo']**2)
+
+                    xlo, xhi = (mean - sigma_lo, mean + sigma_hi)
+                    ylo, yhi = self.plotter.yrange
+
+                    rect = matplotlib.patches.Rectangle((xlo, ylo), width=xhi - xlo, height=yhi - ylo,
+                                                       alpha=self.alpha, color=self.color)
+                    self.plotter.ax.add_patch(rect)
+                else: # 'y' in self.item
```
Comment (pluegh) [inline python/eos/plot/plotter.py:816]: `if 'y' in self.item`

**Review by pluegh — CHANGES_REQUESTED** [review]
Looking good except for a single typo. And I would avoid `else` altogether when all cases that are tested against are known.

File: python/eos/plot/plotter.py [hunk 5]
```diff
@@ -671,6 +672,181 @@ def plot(self):
...
+                    rho    = np.array(constraint['correlations'])[xidx, yidx]
+
+                sigmas = [1.0, 2.0, 3.0]
+                alphas = np.linspace(0.0, self.alpha, len(sigmas) + 1)[1:]
+                for sigma, alpha in zip(sorted(sigmas, reverse=True), alphas):
+                    xwidth = 2.0 * xsigma * np.sqrt(1.0 + rho) * sigma
+                    ywidth = 2.0 * ysigma * np.sqrt(1.0 - rho) * sigma
+
+                    ellipse = matplotlib.patches.Ellipse((xmean, ymean), width=xwidth, height=ywidth, angle=45,
+                                          alpha=alpha, color=self.color, linewidth=1, fill=True)
```
Comment (mreboud) [inline python/eos/plot/plotter.py:784]: I don't understand your ellipse, why is the angle set at 45? Shouldn't this angle depend on the correlation?

**Review by mreboud — CHANGES_REQUESTED** [review]

File: python/eos/plot/plotter.py [hunk 6]
```diff
@@ -671,6 +672,181 @@ def plot(self):
...
+
+                    xidx = observables.index(xobservable)
+                    yidx = observables.index(yobservable)
+
+                    means  = np.array(constraint['means'])
+                    xmean  = means[xidx]
+                    ymean  = means[yidx]
+
+                    sigmas = np.sqrt(
+                          (np.array(constraint['sigma-stat-hi'])**2 + np.array(constraint['sigma-stat-lo'])**2) / 4.0
```
[the remark this answers is not in the document]
Comment (dvandyk) [reply python/eos/plot/plotter.py:769]: Discussed this with Meril. We will use the square of sum of absolute values for the time being.

File: python/eos/plot/plotter.py (hunk 5 above)
Comment (dvandyk) [reply python/eos/plot/plotter.py:784]: I will investigate. Might have implemented the matplotlib example incorrectly.

Comment (dvandyk) [reply python/eos/plot/plotter.py:785]: I made a mistake. The various transformations do not commute. Fixed.

**Pull request merged**